tb_plugin/torch_tb_profiler/profiler/gpu_metrics_parser.py

```diff
@@ -0,0 +1,163 @@
+
+                for i_bucket in range(buckets):
+                    start_time = steps_start_time + i_bucket * bucket_size
+                    self.gpu_util_buckets[gpu_id].append((start_time, gpu_utilization_timeline[-1][i_bucket]))
+                start_time = steps_start_time + buckets * bucket_size
+                self.gpu_util_buckets[gpu_id].append((start_time, 0))
+
+        self.kernel_ranges_per_device = None  # Release memory.
+
+    def calculate_approximated_sm_efficency(self, steps_start_time, steps_end_time):
+        def calculate_avg(approximated_sm_efficency_ranges, total_dur):
+            total_weighted_sm_efficiency = 0.0
+            for r in approximated_sm_efficency_ranges:
+                dur = r[0][1] - r[0][0]
+                total_weighted_sm_efficiency += r[1] * dur
+            avg_approximated_sm_efficency = total_weighted_sm_efficiency / total_dur
+            return avg_approximated_sm_efficency
+
+        total_dur = steps_end_time - steps_start_time
+        for gpu_id in self.gpu_ids:
+            blocks_per_sm_ranges = self.blocks_per_sm_per_device[gpu_id]
+            approximated_sm_efficency_ranges = merge_ranges_with_value(blocks_per_sm_ranges)
+            avg_approximated_sm_efficency = calculate_avg(approximated_sm_efficency_ranges, total_dur)
+            self.avg_approximated_sm_efficency_per_device[gpu_id] = avg_approximated_sm_efficency
+
+            if avg_approximated_sm_efficency > 0.0:
+                self.approximated_sm_efficency_ranges[gpu_id] = approximated_sm_efficency_ranges
+
+        self.blocks_per_sm_per_device = None  # Release memory.
+
+    # Weighted average. Weighted by kernel's time duration.
+    def calculate_occupancy(self):
+        for gpu_id in self.gpu_ids:
+            occupancys_on_a_device = self.occupancy_per_device[gpu_id]
+            total_time = 0
+            total_occupancy = 0.0
+            for r in occupancys_on_a_device:
+                dur = r[0][1] - r[0][0]
+                total_occupancy += r[1] * dur
+                total_time += dur
+            avg_occupancy = total_occupancy / total_time
+            self.avg_occupancy_per_device[gpu_id] = avg_occupancy
+
+    def parse_events(self, events, steps_start_time, steps_end_time):
+        logger.debug("GPU Metrics, parse events")
+        for event in events:
+            self.parse_event(event)
+
+        self.calculate_gpu_utilization(steps_start_time, steps_end_time)
+        self.calculate_approximated_sm_efficency(steps_start_time, steps_end_time)
+        self.calculate_occupancy()
+
+    def parse_event(self, event):
+        ts = event.ts
+        dur = event.duration
+        evt_type = event.type
+        if evt_type == EventTypes.KERNEL:
+            gpu_id = event.args.get("device", None)
+            if gpu_id is not None:
+                if gpu_id not in self.gpu_ids:
+                    self.kernel_ranges_per_device.extend([[] for _ in range(gpu_id + 1 - len(self.gpu_ids))])
```

Comment:
what if the gpu_id is in reverse order in events?

Reply:
Thanks. All changed to length 64.